### ISSUE TITLE
beacons early exit when the sliver server cannot be reach for a period of time. 

### ISSUE THREAD

**Issue by HellooTiger** [BishopFox/sliver#1253]
**Describe the bug**
A beacon will exit unexpectedly after the sliver server is taken down for a while.  Is there anyway to config the beacons that they never exit by itself regardless whether the sliver server cannot be reached for a period of time. Sometimes, the server is intentionally taken download or cannot be online for one day or two.

**Comment by moloch--** [comment]
You should be able to adjust this behavior by setting `--max-errors` and  `--reconnect`. Where `--max-errors` is the maximum number of connection errors before the implant exits and `--reconnect` is the amount of time between connection attempts if the previous connection attempt failed (not to be confused with check-ins).

By default `--max-errors` is 1000 and `--reconnect` is set to 60 seconds; meaning the implant should attempt to reconnect every 60 seconds if the previous connection failed up to 1000 times which is about 16 hours.

While we don't have an indefinite option, you can set `--max-errors` to a very large value to get nearly indefinite reconnects.

**Comment by xinhuang** [comment]
I have the same question earlier.

I've checked the source code about the `--max-errors` option and it seems that on the surface, i.e. the command line option it's an integer, as here shows: https://github.com/BishopFox/sliver/blob/e030b7cdd0e713cc522d6fe024d08a2f39a45cfb/client/command/generate/generate.go#L267

It becomes an unsigned integer in implant, client, etc. As shown in here:
https://github.com/BishopFox/sliver/blob/e030b7cdd0e713cc522d6fe024d08a2f39a45cfb/server/db/models/implant.go#L94

Not sure whether they are connected, but it might be better to unify the type and add an infinite retry option. (Bad for opsec for sure, good for HackTheBox and other labs ;)

**Comment by HellooTiger** [comment]
> By default `--max-errors` is 1000 and `--reconnect` is set to 60 seconds; meaning the implant should attempt to reconnect every 60 seconds if the previous connection failed up to 1000 times which is about 16 hours.

That's really good. I even found that the '--reconnect ' parameter can be adjusted by using "reconfig -r" command when a beacon checks in. I wonder if it is possible to display the '--reconnect ' parameter by using "info" command when manipulating a beacon.  It would be more informative with the feature when manipulating a beacon.  For now, only "beacon interval" and "beacon jitter" are shown. 

Another wanted feature  is to add an option for 'reconnect jitter' just as beacon jitter. This would make the network communication of beacons reconnecting unpredictable and irregularly when  the sliver server is intentionally taken down for a couples of days or weeks to avoid being hunted by blue team.

**Comment by HellooTiger** [comment]
> It becomes an unsigned integer in implant, client, etc. As shown in here:

It's interesting. Maybe the checking logic of --max-errors can be removed to simulate a infinite loop when implants are being customized.

**Comment by moloch--** [comment]
We certainly could add an indefinite, but this seems risky from a professional red team standpoint. Jitter could also be added, I don't think it would be much work but the implant should (hopefully) be checking in to the server most of the time and the standard jitter settings will apply.

**Comment by pythonSB** [comment]
> 我们当然可以添加一个不确定的，但从专业红队的角度来看这似乎很冒险。还可以添加抖动，我认为这不会花费太多工作量，但植入物应该（希望）大多数时间都会检查服务器，并且标准抖动设置将适用。

I think from any perspective, you should not exit the process after the timeout, which will cause the loot to be lost.

**Comment by Nm1ss** [comment]
Does the max-error counter reset after a successful connection?

**Comment by moloch--** [comment]
Yes, it is reset https://github.com/BishopFox/sliver/blob/master/implant/sliver/runner/runner.go#L236
